controller/src/main/java/io/pravega/controller/server/bucket/BucketManagerLeader.java

```diff
@@ -0,0 +1,192 @@
+        this.bucketDistributor = bucketDistributor;
+        this.serviceType = serviceType;
+    }
+
+    /**
+     * Suspend the leader thread.
+     */
+    public void suspend() {
+        suspended.set(true);
+    }
+
+    /**
+     * Resume the suspended leader thread.
+     */
+    public void resume() {
+        if (suspended.compareAndSet(true, false)) {
+            suspendMonitor.release();
+        }
+    }
+
+    @Override
+    public void takeLeadership(CuratorFramework client) throws Exception {
+        log.info("{}: Obtained leadership to monitor the controller to buckets mapping.", serviceType);
+
+        // Attempt a distribution whenever leadership is obtained to ensure no Controllers events are missed.
+        controllerChange.release();
+
+        // Start cluster monitor.
+        pravegaServiceCluster = new ClusterZKImpl(client, ClusterType.CONTROLLER);
+
+        // Add listener to track controller changes on the monitored pravega cluster.
+        pravegaServiceCluster.addListener((type, controller) -> {
+            switch (type) {
+                case HOST_ADDED:
+                case HOST_REMOVED:
+                    // We don't keep track of the controllers and we always query for the entire set from the cluster
+                    // when changes occur. This is to avoid any inconsistencies if we miss any notifications.
+                    log.info("{}: Received event: {} for host: {}. Wake up leader for distribution", serviceType, type, controller);
+                    controllerChange.release();
+                    break;
+                case ERROR:
+                    // This event should be due to ZK connection errors and would have been received by the monitor too,
+                    // hence not handling it explicitly here.
+                    log.info("{}: Received error event when monitoring the pravega host cluster, ignoring...", serviceType);
+                    break;
+            }
+        });
+
+        // Keep looping here as long as possible to stay as the leader and exclusively monitor the pravega cluster.
+        while (true) {
+            try {
+                if (suspended.get()) {
+                    log.info("{}: Monitor is suspended, waiting for notification to resume.", serviceType);
+                    suspendMonitor.acquire();
+                    log.info("{}: Resuming monitor.", serviceType);
+                }
+
+                controllerChange.acquire();
+                log.info("{}: Received distribute buckets.", serviceType);
+
+                // Wait here until distribution can be performed.
+                waitForReDistribute();
+
+                // Clear all events that has been received until this point since this will be included in the current
+                // distribution operation.
+                controllerChange.drainPermits();
+                triggerDistribution();
+            } catch (InterruptedException e) {
```

Comment:
The exception caught here is being thrown using `throw e;` and it is re-interrupted at org.apache.curator.framework.recipes.leader.LeaderSelector at line no 493.
Here is the code snippet 
` catch ( InterruptedException ignore )
        {
            Thread.currentThread().interrupt();
        }`

Reply:
More or less everywhere we use `Exceptions.handleInterupted(_)` which provides uniformity and makes it obvious that it's being handled correctly. I would prefer using that rather than relying on external code, because it is easier to audit and isn't subject to change if we update our version of curator.

Reply:
Basically this class implements the interface `LeaderSelectorListener (org.apache.curator.framework.recipes.leader)` which is of curator. Also this method is not directly called by us. Its called by `LeaderSelector(org.apache.curator.framework.recipes.leader)` class which is also of curator framework. We have to pass this listener as a argument while initializing `LeaderSelector `class i.e ` new LeaderSelector(client, zkpath, listener)`.

Reply:
Re-interrupt the thread as suggested.